libflux/src/semantic/parser/mod.rs

```diff
@@ -0,0 +1,1506 @@
+            "forall" => TokenType::FORALL, 
+            "where" => TokenType::WHERE, 
+            "int" => TokenType::INT, 
+            "float" => TokenType::FLOAT, 
+            "string" => TokenType::STRING, 
+            "bool" => TokenType::BOOL, 
+            "uint" => TokenType::UINT, 
+            "duration" => TokenType::DURATION, 
+            "time" => TokenType::TIME, 
+            "regexp" => TokenType::REGEXP, 
+            _ => TokenType::IDENTIFIER
+        }
+    }
+}
+
+// lex_token recursively lexes characters using StateFn
+fn lex_token(lexer: &mut Lexer) -> Option<TokenType> {
+    match lexer.next() {
+        Some(letter) if letter.is_alphanumeric() =>  Some(lexer.keyword_or_ident()), 
+        Some(letter) if letter.is_whitespace() => Some(TokenType::WHITESPACE), 
+        Some(letter) if letter == '{' => Some(TokenType::LEFTCURLYBRAC), 
+        Some(letter) if letter == '}' => Some(TokenType::RIGHTCURLYBRAC), 
+        Some(letter) if letter == '[' => Some(TokenType::LEFTSQUAREBRAC), 
+        Some(letter) if letter == ']' => Some(TokenType::RIGHTSQUAREBRAC), 
+        Some(letter) if letter == '(' => Some(TokenType::LEFTPAREN), 
+        Some(letter) if letter == ')' => Some(TokenType::RIGHTPAREN), 
+        Some(letter) if letter == '?' => Some(TokenType::QUESTIONMARK), 
+        Some(letter) if letter == ':' => Some(TokenType::COLON), 
+        Some(letter) if letter == ',' => Some(TokenType::COMMA), 
+        Some(letter) if letter == '+' => Some(TokenType::PLUS), 
+        Some(letter) if letter == '|' => Some(TokenType::WITH),
+        Some(letter) if letter == '<' => {
+            match lexer.next() {
+                Some('-') => Some(TokenType::PIPE),
+                _ => Some(TokenType::ERROR), 
+            }
+        },
+        Some(letter) if letter == '-' => {
+            match lexer.next() {
+                Some('>') => Some(TokenType::ARROW), 
+                _ => Some(TokenType::ERROR), 
+            }
+        },
+        Some(_) => Some(TokenType::ERROR), 
+        _ => Some(TokenType::EOF), 
+    }
+}
+
+/** PARSER */
+
+struct Parser<'a> {
+    tokens: Peekable<Iter<'a, Token>>,
+}
+
+// parse_tokens is a testing function for directly parsing a vector of tokens. 
+// If parsing a &str polytype, use the Parse function instead
+// fn parse_tokens(tokens: Vec<Token>) -> Result<PolyType, &'static str> {
+//     let mut parser = Parser{ tokens: tokens.iter().peekable() };
+//     parser.parse_polytype()
+// }
+
+// Parse passes the source text through the Lexer, receives tokens and removes
+// whitespace. Its then initializes parsing and returns a PolyType AST.
+// This is the only function meant to be accessed by the end user. It handles 
+// both lexing and parsing of string polytypes. 
+pub fn Parse(source: &str) -> Result<PolyType, &'static str> {
+    let tokens = Lex(source); 
+    let mut parser = Parser{ tokens: tokens.iter().peekable() }; 
+    parser.parse_polytype()
```

Comment:
One way to refactor this is to keep `parse_tokens` but write `Parse` in terms of `parse_tokens. 

```
pub fn Parse(source: &str) -> Result<PolyType, &'static str> {
    let tokens = Lex(source); 
    parse_tokens(tokens)
```

this way the decoupled nature of the code still exists so we can test things in isolation, while still presenting a simple API to end users of the code.

Reply:
I went ahead and just removed `parse_tokens()`. Should be simple to add it back in if the need arises.